core/utils/utils.go

```diff
@@ -1112,3 +1114,50 @@ func MinKey[U any, T constraints.Ordered](elems []U, key func(U) T) T {
 
 	return min
 }
+
+// ErrorBuffer uses joinedErrors interface to join multiple errors into a single error.
+// This is useful to track the most recent N errors in a service and flush them as a single error.
+
+type ErrorBuffer struct {
+	// buffer is a slice of errors
+	buffer []error
+	// Cap is the maximum number of errors that the buffer can hold.
+	// Exceeding the cap results in discarding the oldest error
+	Cap int
+
+	sync.RWMutex
+}
+
+func (eb *ErrorBuffer) Flush() (err error) {
+	eb.RLock()
+	defer eb.RUnlock()
+	err = errors.Join(eb.buffer...)
+	eb.buffer = nil
+	return
+}
+
+func (eb *ErrorBuffer) Append(incoming error) {
+	eb.Lock()
+	defer eb.Unlock()
+	// if at capacity, drop first element
+	if len(eb.buffer) == eb.Cap && eb.Cap != 0 {
+		eb.buffer = append(eb.buffer[1:], incoming)
+		return
+	}
+	eb.buffer = append(eb.buffer, incoming)
+}
+
+type joinedError interface {
+	Unwrap() []error
+}
+
+// UnwrapError returns a list of underlying errors if passed error implements joinedError or return the err in a single-element list otherwise.
+//
+//nolint:errorlint // error type checks will fail on wrapped errors. Disabled since we are not doing checks on error types.
+func UnwrapError(err error) []error {
+	joined, ok := err.(joinedError)
```

Comment:
WDYT about making this interface anonymous?
```suggestion
// UnwrapError returns a list of underlying errors if passed error implements joinedError or return the err in a single-element list otherwise.
//
//nolint:errorlint // error type checks will fail on wrapped errors. Disabled since we are not doing checks on error types.
func UnwrapError(err error) []error {
	joined, ok := err.(interface { Unwrap() []error })
```
b/c technically implementors are not limited to _joined_ errors.

Reply:
I actually thought of doing so, oddly enough, benchmarks (locally) favored extracting it out over anon. Probably a weird allocation gimmick. Happy to change though, if stdlib is going this way, it might be that my local benchmarks are not accurate :shrug:

Reply:
confirmed again, inline anon is slower

Simple bench
```
func BenchmarkTestBuffer(b *testing.B) {
	var errs []error
	err1 := errors.New("testerror")
	for n := 0; n < b.N; n++ {
		// always record the result of Fib to prevent
		// the compiler eliminating the function call.
		errs = utils.UnwrapError(err1)
	}
	// always store the result to a package level variable
	// so the compiler cannot eliminate the Benchmark itself.
	_ = errs
}
```

Anonymous inline
```
Running tool: /usr/bin/go test -benchmem -run=^$ -bench ^BenchmarkTestBuffer$ github.com/smartcontractkit/chainlink/core/utils

cltest random seed: 1678122878783330221
goos: linux
goarch: amd64
pkg: github.com/smartcontractkit/chainlink/core/utils
cpu: 11th Gen Intel(R) Core(TM) i7-1185G7 @ 3.00GHz
BenchmarkTestBuffer-8   	12333069	       103.4 ns/op	      16 B/op	       1 allocs/op
PASS
ok  	github.com/smartcontractkit/chainlink/core/utils	1.813s
```


Declared type
```
Running tool: /usr/bin/go test -benchmem -run=^$ -bench ^BenchmarkTestBuffer$ github.com/smartcontractkit/chainlink/core/utils

cltest random seed: 1678122957173522065
goos: linux
goarch: amd64
pkg: github.com/smartcontractkit/chainlink/core/utils
cpu: 11th Gen Intel(R) Core(TM) i7-1185G7 @ 3.00GHz
BenchmarkTestBuffer-8   	38731735	        29.06 ns/op	      16 B/op	       1 allocs/op
PASS
ok  	github.com/smartcontractkit/chainlink/core/utils	2.015s
```

Reply:
scratch that, after 10-20 reps, it seems its not consistent